src/Featurizers/TruncatedSVDFeaturizer.h

```diff
@@ -0,0 +1,631 @@
+    typedef typename Eigen::Matrix<Scalar, Eigen::Dynamic, Eigen::Dynamic> DenseMatrix;
+
+    Eigen::Index rank = (_matrix.rows() < _matrix.cols()) ? _matrix.rows() : _matrix.cols();
+
+    // Gaussian Random Matrix for _matrix^T
+    DenseMatrix O(_matrix.rows(), rank);
+    sample_gaussian(O);
+			
+    // Compute Sample Matrix of _matrix^T
+    DenseMatrix Y = _matrix.transpose() * std::move(O);
+			
+    // Orthonormalize Y
+    gram_schmidt(Y);
+			
+    // Range(B) = Range(_matrix^T)
+    DenseMatrix B = _matrix * Y;
+			
+    // Gaussian Random Matrix
+    DenseMatrix P(B.cols(), rank);
+    sample_gaussian(P);
+			
+    // Compute Sample Matrix of B
+    DenseMatrix Z = B * P;
+			
+    // Orthonormalize Z
+    gram_schmidt(Z);
+			
+    // Range(C) = Range(B)
+    DenseMatrix C = Z.transpose() * std::move(B); 
+			
+    Eigen::JacobiSVD<DenseMatrix> svdOfC(C, Eigen::ComputeThinV);
+			
+    // C = USV^T
+    // A = Z * U * S * V^T * Y^T()
+    _singularvalues = svdOfC.singularValues();
+    _singularvectors = std::move(Y) * svdOfC.matrixV();
+
+    return SVDComponentsAnnotationData<TransformedT>(std::move(_singularvalues), std::move(_singularvectors));
+}
+
+// ----------------------------------------------------------------------
+// |
+// |  SVDTransformer
+// |
+// ----------------------------------------------------------------------
+template <typename InputT, typename TransformedT>
+SVDTransformer<InputT, TransformedT>::SVDTransformer(TransformedT singularvalues, TransformedT singularvectors) :
+    //no validation here because same variable has been validated in annotation class
+    _singularvalues(std::move(singularvalues)),
+    _singularvectors(std::move(singularvectors)) {
+}
+
+//TODO:
+// template <typename InputT, typename TransformedT>
+// SVDTransformer<InputT, TransformedT>::SVDTransformer(Archive &ar) :
+//     SVDTransformer(
+//         [&ar](void) {
+//             // Version
+//             std::uint16_t                   majorVersion(Traits<std::uint16_t>::deserialize(ar));
+//             std::uint16_t                   minorVersion(Traits<std::uint16_t>::deserialize(ar));
+
+//             if(majorVersion != 1 || minorVersion != 0)
+//                 throw std::runtime_error("Unsupported archive version");
+
+//             // Data:TODO
+//             TransformedT                   singularvalues(Traits<TransformedT>::deserialize(ar));
+//             TransformedT                   singularvectors(Traits<TransformedT>::deserialize(ar));
+
+//             return SVDTransformer<InputT, TransformedT>(std::move(singularvalues), std::move(singularvectors));
+//         }()
+//     ) {
+// }
+
+template <typename InputT, typename TransformedT>
+bool SVDTransformer<InputT, TransformedT>::operator==(SVDTransformer const &other) const {
+    if ((this->_singularvalues - other._singularvectors).norm() > 0.000001f || (this->_singularvalues - other._singularvectors).norm() > 0.000001f)
```

Comment:
>if ((this->_singularvalues - other._singularvectors).norm() > 0.000001f || (this->_singularvalues - other._singularvectors).norm() > 0.000001f) [](start = 4, length = 143)

I think something is off here.

Reply:
It's a bug. Thanks for pointing that out. Already been corrected and will be in the next update